src/mame/trs/agvision.cpp

```diff
@@ -0,0 +1,417 @@
+
+	Dynamic RAM (4, 16, or 39K) starts at $0000.
+	ROM (2K) starts at $A000 and mirrors up to $BFFF
+	Static RAM starts at $C000 for 128 bytes
+	The PIA data port a is at $FF1C
+	     control port a is at $FF1D
+		    data port b is at $FF1E
+		 control port b is at $FF1F
+
+	The Control Register is at $FF20
+	The SAM is at $FFC0
+
+	PIA:
+		Port A and B make up the keyboard matrix
+		PA7  - RS-232 recieve.
+		CA1  - The horizontal sync from the VDG.
+		CA2  - Unconnected, not verified.
+		CB1  - Modem status, carrier detect.
+		CB2  - Modem control (unverified).
+		IRQA - Connected to the 6809 FIRQ.
+		IRQB - Unconnected, not verified.
+
+	Control Register:
+		Bit 7 - RS-232 transmit.
+		Bit 6 - RS-232 DTR, or activity LED (unsure which).
+		Bit 5 - unconnected (not verified).
+		Bit 3 - VDG GM0.
+		Bit 2 - VDG GM1.
+		Bit 1 - VDG GM2.
+		Bit 0 - VDG A* / G.
+
+	Reverse engineering used with permission of George Phillips of the
+	trs80gp emulator.
+
+***************************************************************************/
+
+#include "emu.h"
+
+#include "cpu/m6809/m6809.h"
+#include "machine/6821pia.h"
+#include "machine/ram.h"
+#include "screen.h"
+#include "6883sam.h"
+#include "video/mc6847.h"
+#include "bus/rs232/rs232.h"
+
+//-------------------------------------------------
+//  MACROS / CONSTANTS
+//-------------------------------------------------
+
+#define MAINCPU_TAG     "maincpu"
+#define PIA0_TAG        "pia0"
+#define SCREEN_TAG      "screen"
+#define RS232_TAG       "rs232"
+#define SAM_TAG         "sam"
+#define VDG_TAG         "vdg"
+
+//-------------------------------------------------
+//  TYPE DEFINITIONS
+//-------------------------------------------------
+
+class agvision_state : public driver_device
+{
+public:
+	agvision_state(const machine_config &mconfig, device_type type, const char *tag)
+		: driver_device(mconfig, type, tag)
+		, m_maincpu(*this, MAINCPU_TAG)
+		, m_screen(*this, SCREEN_TAG)
+		, m_ram(*this, RAM_TAG)
+		, m_pia_0(*this, PIA0_TAG)
+		, m_sam(*this, SAM_TAG)
+		, m_vdg(*this, VDG_TAG)
+		, m_rs232(*this, RS232_TAG)
+		, m_keyboard(*this, "row%u", 0)
+	{}
+
+	void agvision(machine_config &config);
+	uint8_t sam_read(offs_t offset);
+	DECLARE_INPUT_CHANGED_MEMBER(hang_up);
+
+protected:
+	virtual void device_start() override;
+
+	required_device<cpu_device> m_maincpu;
+	required_device<screen_device> m_screen;
+	required_device<ram_device> m_ram;
+	required_device<pia6821_device> m_pia_0;
+	required_device<sam6883_device> m_sam;
+	required_device<mc6847_base_device> m_vdg;
+	required_device<rs232_port_device> m_rs232;
+	required_ioport_array<7> m_keyboard;
+
+	emu_timer * m_timer; // Carrier Detect Timer
+
+	void agvision_mem(address_map &map);
+	void agvision_rom(address_map &map);
+	void agvision_static_ram(address_map &map);
+	void agvision_io0(address_map &map);
+	void agvision_io1(address_map &map);
+	void agvision_boot(address_map &map);
+    void ff20_write(offs_t offset, uint8_t data);
```

Comment:
Spaces instead of tab :)

Reply:
Fixed.

Reply:
Sorry, I think you reindented everything with spaces instead of tabs due to my brain fart making me write that sentence backwards. I only meant to say that that one line (line 108) was using 4 spaces when it should have been 1 tab like the rest. Mea culpa.

Reply:
You can just run srcclean on the file to fix it I think if you have the MAME tools built.